azure-kusto-data/azure/kusto/data/security.py

```diff
@@ -198,15 +253,81 @@ def _acquire_authorization_header(self) -> str:
         elif self.authentication_method is AuthenticationMethod.aad_application_certificate:
             token = self.auth_context.acquire_token_with_client_certificate(self.kusto_uri, self.client_id, self.certificate, self.thumbprint)
         else:
-            raise KustoClientError("Please choose authentication method from azure.kusto.data.security.AuthenticationMethod")
+            raise self._authentication_method_missing_exception()
 
         return _get_header_from_dict(token)
 
+    async def _acquire_authorization_header_async(self) -> str:
+        """Async version of self._acquire_authorization_header"""
+        if self.authentication_method is AuthenticationMethod.aad_token:
+            return _get_header("Bearer", self.token)
+
+        # Obtain token from MSI endpoint
+        if self.authentication_method == AuthenticationMethod.aad_msi:
+            token = await self.get_token_from_msi_async()
+            return _get_header_from_dict(token)
+
+        refresh_token = None
+
+        if self.authentication_method == AuthenticationMethod.az_cli_profile:
+            stored_token = await _get_azure_cli_auth_token_async()
+            self._set_from_stored_token(stored_token)
+            refresh_token = stored_token[TokenResponseFields.REFRESH_TOKEN]
+
+        if self.auth_context is None:
+            self.auth_context = AuthenticationContext(self.authority_uri)
+
+        if refresh_token is not None:
+            token = await sync_to_async(self.auth_context.acquire_token_with_refresh_token)(refresh_token, self.client_id, self.kusto_uri)
```

Comment:
I'm not sure this is safe. Adal's AuthContext was not built with multi-threading in mind.

Reply:
From asgi docs:
"If the outermost program is async (i.e. SyncToAsync is outermost), then
    this will be a dedicated single sub-thread that all sync code runs in,
    one after the other"